consolidate inferior-process sending: prefix arg use, escape region

ilohmar opened the pull request:
This removes the `-and-follow' command variants.  When working in a
source buffer, it is common to send many parts to the inferior process,
staying in the source buffer.  Switching to the inferior process is the
exception from that, and a wonderful use of the optional prefix arg.

Also for a more DWIM behavior, when a region is active, send that
instead of the current line.

The sending itself now escapes the contents, permitting multi-line
sends.

It seems https://github.com/museoa/bqn-mode/pull/7 had some code for
that which was not merged.

The current escaping code is ugly and wasteful, but easy to understand.

ilohmar committed:
[improvement] Consolidate inferior-process sending

The typical use case for inferior-process sending is to send many
snippets of code to the inferior process while staying in the source
buffer.  Switching to the inferior process is the exception rather than
the rule; this naturally justifies the optional prefix argument.

This change obviates the `-and-follow` variants, therefore they will be
removed.

In the course of these changes, avoid user commands to operate on
buffers, and when switching buffers, resort to commands that respect
user customizations.

Also for a more "do-what-I-mean" behavior, when a region is active, send
that region instead of the current line.

Being now escaped, the sending now works with multiple lines.

ilohmar commented:
This had trivial but really annoying merge conflicts.. will repair it first.

AndersonTorres commented:
The code(s) in #7 were not merged mainly because I couldn't understand it :)

Now I am far from my "workstation", but I will rebase this at night.

ilohmar commented:
Great.  If you're only rebasing because I did not follow commit conventions, please let me know, I will try to follow them.

ilohmar committed:
[improvement] Do not insert source buffer parts to send them

This is the typical usage, especially important when sending larger
chunks (like a whole buffer).  It's even more important now that the
content uses the ")escape" construction.

Avoid sending initial newline, because otherwise the evaluation of an
empty line will be triggered, resulting in a spurious prompt that
destroys output alignment.

ilohmar commented:
I have a series of follow-up changes based on this PR, including several fixes, cleanup etc.  If you want to have a look, they are at https://github.com/ilohmar/bqn-mode/tree/comint.

AndersonTorres commented:
There are no commit conventions yet, except my personal feeling of aesthetics :)

AndersonTorres commented on bqn-mode.el:
```diff
@@ -519,47 +519,56 @@ If it doesn't exist, create and return it; else, return the existing one."
 
-(defun bqn-comint-process-execute-region (start end &optional dont-follow)
+(defun bqn-comint--escape (str)
+  "Targeting CBQN."
```
A better documentation would be good, even if this is not intended to be seen by the end user.
Especially explaining why this is needed.

ilohmar replied:
I changed it to a comment instead, internal functions should generally not use docstrings.

AndersonTorres commented on bqn-mode.el:
```diff
@@ -519,47 +519,56 @@ If it doesn't exist, create and return it; else, return the existing one."
-    (error
-     (concat "Attempt to send empty region to "
-             bqn-comint-*process-buffer-name*)))
+    (error "Attempt to send empty region to %s" bqn-comint--process-name))
```
wow, does it work? O.o

ilohmar replied:
Why wouldn't it?

AndersonTorres commented on bqn-mode.el:
```diff
@@ -519,47 +519,56 @@ If it doesn't exist, create and return it; else, return the existing one."
+      (select-window (display-buffer pbuf)))))
+
+(defun bqn-comint-process-execute-line (&optional arg)
+  "Send the active region, else the current line to the BQN process."
```
It looks too overloaded to me. 
I would at least rename this to ` bqn-comint-process-execute-line-or-region`

ilohmar replied:
Done.

AndersonTorres commented on bqn-mode.el:
```diff
@@ -519,47 +519,56 @@ If it doesn't exist, create and return it; else, return the existing one."
+(defun bqn-comint-process-execute-line (&optional arg)
+  "Send the active region, else the current line to the BQN process."
+  (interactive "P")
+  (cond
```
Isn't `if` better than `cond` here? There is only one condition being tested, namely `(use-region-p)`.

ilohmar replied:
Done

AndersonTorres commented:
First commit message:

```
[improvement] Consolidate inferior-process sending

The typical use case for inferior-process sending is to send many snippets of code to the inferior process while staying in the source buffer.  Switching to the inferior process is the exception rather than the rule; this naturally justifies the optional prefix argument.

This change obviates the `-and-follow` variants, therefore they will be removed.

In the course of these changes, avoid user commands to operate on buffers, and when switching buffers, resort to commands that respect user customizations.

Also for a more "do-what-I-mean" behavior, when a region is active, send that region instead of the current line.

Being now escaped, the sending now works with multiple lines.
```

AndersonTorres commented:
Second message:

```
[improvement] Send source buffer parts to inferior process without inserting

This is the typical usage, especially important when sending larger chunks (like a whole buffer).  It's even more important now that the content uses the ")escape" construction.

Avoid sending initial newline, because otherwise the evaluation of an empty line will be triggered, resuling in a spurious prompt that destroys output alignment.
```

ilohmar commented:
> Second message:
> 
> ```
> [improvement] Send source buffer parts to inferior process without inserting
> 
> This is the typical usage, especially important when sending larger chunks (like a whole buffer).  It's even more important now that the content uses the ")escape" construction.
> 
> Avoid sending initial newline, because otherwise the evaluation of an empty line will be triggered, resuling in a spurious prompt that destroys output alignment.
> ```

This makes the summary line far too long, I cut it short.  Please consider my questions in the other PR whether you want to insist on using the uncommon prefixes.

ilohmar commented:
Ok, I have been thru a rebase churn and still managed to wreck this (not all changes are included, although I claimed they were).  This has made me pause and reflect what I want to do.

My heart is not really in the incremental-changes-via-PR business here: I have more or less a rewrite of some parts of bqn-mode in mind (or already started it), and an overhaul of most others.  Any review cycle implies rebasing my local code; that has already diverted all my attention from further plans I had.  Also, I respectfully disagree with many of your conventions as a matter of principle.  Overall, this feels like too much busy-work for my spare time, and would burn your time as well.

I will be away for a few days soon.  If I continue to work on the mode after that, it will most likely be simply on my fork.

AndersonTorres commented:
> > Second message:
> > 
> > ```
> > [improvement] Send source buffer parts to inferior process without inserting
> > 
> > This is the typical usage, especially important when sending larger chunks (like a whole buffer).  It's even more important now that the content uses the ")escape" construction.
> > 
> > Avoid sending initial newline, because otherwise the evaluation of an empty line will be triggered, resuling in a spurious prompt that destroys output alignment.
> > ```
> 
> This makes the summary line far too long

I was focusing more on the content, and I forgot to break the lines.

AndersonTorres commented:
I feel your decision was a bit hasty. 

Overall your code is too good to be forgotten. My main concern is about the wording of commit messages.

Please, take your time and ponder about the contributions. I am truly excited to see more of them!

ilohmar commented:
Hi, thanks for your kind words.  I'm really sorry about the way this turned out, it just quickly got tedious -- NOT saying this is your fault, more like the nature of the situation.  I was just playing around with bqn and did that, naturally, from within Emacs.  Trying to fix certain things in bqn-mode then escalated into rewriting parts.

I simply feel that the incremental-PR approach is not well-suited in that context.  For the same reason, I am not really interested in the commit messages at this point: while I try to keep commits and branches self-contained, those are just somewhat artificial in-between steps to a target state.

Ultimately, this is just a matter of how I want to spend my time.  With that out of the way, I will just write the stuff I have in mind, and maybe we can arrange a merge of parts at some point?  I am well aware that this will generally be more difficult, but it's the (really selfish) choice I am making.

ilohmar commented:
FWIW, I think I am "done" (for now) with the comint part, see the commits here: https://github.com/ilohmar/bqn-mode/compare/museoa:bqn-mode:trunk...comint

Clarifications to commit messages are always good, although (for the above reasons) I care relatively little in this case.  Maybe have a look at the changes if you have the time, I am open to discuss things (sometime next week).  BTW, I find it very helpful to do actual *merges* for PRs, especially to clarify such things as "areas of work".  Eg, for the comint branch, earlier and later commits may be related and it makes little sense to fix an earlier one independently --- it's the whole branch that matters.

AndersonTorres commented:
> I simply feel that the incremental-PR approach is not well-suited in that context. 

The github interface in certain sense forces this type of workflow. However I don't think it is an unsurmounable drawback. Personally I just treat PRs as "send patches by e-mail" using an awkward interface.

> For the same reason, I am not really interested in the commit messages at this point: while I try to keep commits and branches self-contained, those are just somewhat artificial in-between steps to a target state.

Keeping the patches/PRs self-contained is sufficient.
After all, Elisp code naturally tends to huge patches when compared to other, more terse languages.

> Ultimately, this is just a matter of how I want to spend my time. With that out of the way, I will just write the stuff I have in mind, and maybe we can arrange a merge of parts at some point? I am well aware that this will generally be more difficult, but it's the (really selfish) choice I am making.

This is OK to me. I should focus more on code than in fancy writing of commit messages, blame on me.

AndersonTorres commented:
New PR: https://github.com/museoa/bqn-mode/pull/44
